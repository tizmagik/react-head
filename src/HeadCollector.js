import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Provider } from './headTagsContext';

export default class HeadCollector extends Component {
  static propTypes = {
    headTags: PropTypes.array.isRequired, // eslint-disable-line react/forbid-prop-types
    children: PropTypes.node.isRequired,
  };

  headTags = {
    add: tag => this.props.headTags.push(tag),
  };

  render() {
    return <Provider value={this.headTags}>{React.Children.only(this.props.children)}</Provider>;
  }
}
