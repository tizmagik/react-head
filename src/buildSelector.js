const buildSelector = obj =>
  Object.keys(obj)
    .map(k => `[${k}="${obj[k]}"]`)
    .join('');

export default buildSelector;
