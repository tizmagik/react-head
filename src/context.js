import * as React from 'react';

export const { Consumer, Provider } = React.createContext(null);
