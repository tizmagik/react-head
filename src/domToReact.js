import React from 'react';

/**
 * inspired by https://github.com/remarkablemark/html-react-parser/blob/master/lib/dom-to-react.js
 * but a very stripped down version
 * - no support for
 *  - replace
 *  - transforming attribs (any that are needed should be hard coded here)
 *  - style props (not needed in head)
 * 
 * Will work with script tags however its NOT FOR INJECTING CODE - use this for things like <Script type="application/ld+json" dangerouslySetInnerHTML={{ __html: JSON.stringify(schemaOrg) }}/>
 */
const domToReact =(nodes=[], options={}) => {

  const library = options.library || React;
  const { createElement } = library;

  const result = []

  const len = nodes.length

  for (let i = 0; i < len; i+=1) {
    let node = nodes[i];

    let children = null;
    const props = {...node.attribs}

    switch (node.type) {
      case 'script':
      case 'style':
        // prevent text in <script> or <style> from being escaped
        // https://reactjs.org/docs/dom-elements.html#dangerouslysetinnerhtml
        if (node.children[0]) {
          props.dangerouslySetInnerHTML = {
            __html: node.children[0].data
          };
        }
        break;

      case 'tag':
        if (node.children && node.children.length) {
          // continue recursion of creating React elements (if applicable)
          children = domToReact(node.children, options);
        }
        break;

      // skip all other cases (e.g., comment, text)
      default:
        node = null
    }
    
    if (node) {
      // set "key" prop for sibling elements
      // https://fb.me/react-warning-keys
      if (len > 1) {
        props.key = i;
      }

      result.push(createElement(node.name, props, children));
    }
  }

  return result

}
export default domToReact