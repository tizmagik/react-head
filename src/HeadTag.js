import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import buildSelector from './buildSelector';

export default class HeadTag extends Component {
  static contextTypes = {
    reactHeadTags: PropTypes.object,
  };

  static propTypes = {
    tag: PropTypes.string,
  };

  static defaultProps = {
    tag: 'meta',
  };

  state = {
    canUseDOM: false,
  };

  componentDidMount() {
    // eslint-disable-next-line react/no-did-mount-set-state
    this.setState({ canUseDOM: true });

    const { tag, children, ...rest } = this.props; // eslint-disable-line react/prop-types
    const ssrTags = document.head.querySelector(`${tag}${buildSelector(rest)}[data-rh=""]`);

    /* istanbul ignore else */
    if (ssrTags) {
      ssrTags.remove();
    }
  }

  render() {
    const { tag: Tag, ...rest } = this.props;

    if (this.state.canUseDOM) {
      const Comp = <Tag {...rest} />;
      return ReactDOM.createPortal(Comp, document.head);
    }

    // on client we don't require HeadCollector
    if (this.context.reactHeadTags) {
      const ServerComp = <Tag data-rh="" {...rest} />;
      this.context.reactHeadTags.add(ServerComp);
    }

    return null;
  }
}
