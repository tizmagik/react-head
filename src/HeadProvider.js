import * as React from 'react';
import { Provider } from './context';

const cascadingTags = ['title', 'meta'];

export default class HeadProvider extends React.Component {
  indices = new Map();

  state = {
    addClientTag: (tag, name) => {
      // consider only cascading tags
      if (cascadingTags.indexOf(tag) !== -1) {
        this.setState(state => {
          const names = state[tag] || [];
          return { [tag]: [...names, name] };
        });
        // track indices synchronously
        const { indices } = this;
        const index = indices.has(tag) ? indices.get(tag) + 1 : 0;
        indices.set(tag, index);
        return index;
      }
      return -1;
    },

    shouldRenderTag: (tag, index) => {
      if (cascadingTags.indexOf(tag) !== -1) {
        const names = this.state[tag]; // eslint-disable-line react/destructuring-assignment
        // check if the tag is the last one of similar
        return names && names.lastIndexOf(names[index]) === index;
      }
      return true;
    },

    removeClientTag: (tag, index) => {
      this.setState(state => {
        const names = state[tag];
        if (names) {
          names[index] = null;
          return { [tag]: names };
        }
        return null;
      });
    },

    addServerTag: tagNode => {
      const { headTags = [] } = this.props;
      // tweak only cascading tags
      if (cascadingTags.indexOf(tagNode.type) !== -1) {
        const index = headTags.findIndex(prev => {
          const prevName = prev.props.name || prev.props.property;
          const nextName = tagNode.props.name || tagNode.props.property;
          return prev.type === tagNode.type && prevName === nextName;
        });
        if (index !== -1) {
          headTags.splice(index, 1);
        }
      }
      headTags.push(tagNode);
    },
  };

  componentDidMount() {
    const ssrTags = document.head.querySelectorAll(`[data-rh=""]`);
    // `forEach` on `NodeList` is not supported in Googlebot, so use a workaround
    Array.prototype.forEach.call(ssrTags, ssrTag =>
      ssrTag.parentNode.removeChild(ssrTag)
    );
  }

  render() {
    const { headTags, children } = this.props;

    if (typeof window === 'undefined' && Array.isArray(headTags) === false) {
      throw Error(
        'headTags array should be passed to <HeadProvider /> in node'
      );
    }

    return <Provider value={this.state}>{children}</Provider>;
  }
}
