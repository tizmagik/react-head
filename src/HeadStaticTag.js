import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Consumer } from './context';
import domToReact from "./domToReact"

export default class HeadStaticTag extends React.Component {
  state = {
    canUseDOM: false,
  };

  componentDidMount() {
    this.setState({ canUseDOM: true });
  }

  render() {
    const { canUseDOM } = this.state;

    return (
      <Consumer>
        {api => {
          if (api == null) {
            throw Error('<HeadProvider /> should be in the tree');
          }

          const { json } = this.props
          const components = domToReact(json)
          if (canUseDOM) {
            return ReactDOM.createPortal(components, document.head);
          }
          api.addServerComponents(components);
          return null;
        }}
      </Consumer>
    );
  }
}
