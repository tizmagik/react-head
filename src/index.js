import React from 'react';
import HeadTag from './HeadTag';

export const Title = props => <HeadTag tag="title" {...props} />;

export const Style = props => <HeadTag tag="style" {...props} />;

export const Meta = props => <HeadTag tag="meta" {...props} />;

export const Link = props => <HeadTag tag="link" {...props} />;

export { default as HeadCollector } from './HeadCollector';

export default HeadTag;
