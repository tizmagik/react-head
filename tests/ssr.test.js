/**
 * @jest-environment node
 */

import * as React from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import { HeadProvider, Title, Style, Meta, Link } from '../src';

test('renders nothing and adds tags to headTags context array', () => {
  const arr = [];
  const markup = renderToStaticMarkup(
    <HeadProvider headTags={arr}>
      <div>
        Yes render
        <Title>Title</Title>
        <Style>{`body {}`}</Style>
        <Link href="index.css" />
        <Meta charSet="utf-8" />
      </div>
    </HeadProvider>
  );
  expect(markup).toMatchSnapshot();
  expect(arr).toMatchSnapshot();
});

test('renders only the last title', () => {
  const arr = [];
  renderToStaticMarkup(
    <HeadProvider headTags={arr}>
      <div>
        <Title>Title 1</Title>
      </div>
      <div>
        <Title>Title 2</Title>
      </div>
      <div>
        <Title>Title 3</Title>
      </div>
    </HeadProvider>
  );
  expect(arr).toMatchSnapshot();
});

test('renders only the last meta with the same name', () => {
  const arr = [];
  renderToStaticMarkup(
    <HeadProvider headTags={arr}>
      <Meta>Meta 1</Meta>
      <Meta>Meta 2</Meta>
      <Meta name="1">Meta 3</Meta>
      <Meta name="2">Meta 4</Meta>
      <Meta name="3">Meta 5</Meta>
      <Meta name="2">Meta 6</Meta>
    </HeadProvider>
  );
  expect(arr).toMatchSnapshot();
});

test('renders only last meta with the same property', () => {
  const arr = [];
  renderToStaticMarkup(
    <HeadProvider headTags={arr}>
      <Meta property="name1">Meta 1</Meta>
      <Meta property="name1">Meta 2</Meta>
      <Meta property="name2">Meta 3</Meta>
      <Meta name="name2">Meta 4</Meta>
      <Meta name="name3">Meta 5</Meta>
      <Meta property="name3">Meta 6</Meta>
    </HeadProvider>
  );
  expect(arr).toMatchSnapshot();
});

test('fails if headTags is not passed to <HeadProvider />', () => {
  expect(() => {
    renderToStaticMarkup(
      <HeadProvider>
        <Style>{`body {}`}</Style>
      </HeadProvider>
    );
  }).toThrowError(/headTags array should be passed/);
});

test('fails if headTags is not an array', () => {
  expect(() => {
    renderToStaticMarkup(
      <HeadProvider headTags={{}}>
        <Style>{`body {}`}</Style>
      </HeadProvider>
    );
  }).toThrowError(/headTags array should be passed/);
});

test('throw error if head tag is rendered without HeadProvider', () => {
  const errorFn = jest.spyOn(console, 'error').mockImplementation(() => {});
  expect(() => {
    renderToStaticMarkup(<Style>{`body {}`}</Style>);
  }).toThrowError(/<HeadProvider \/> should be in the tree/);
  errorFn.mockRestore();
});
