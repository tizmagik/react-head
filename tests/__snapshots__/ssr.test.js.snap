// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders nothing and adds tags to headTags context array 1`] = `"<div>Yes render</div>"`;

exports[`renders nothing and adds tags to headTags context array 2`] = `
Array [
  <title
    data-rh=""
  >
    Title
  </title>,
  <style
    data-rh=""
  >
    body {}
  </style>,
  <link
    data-rh=""
    href="index.css"
  />,
  <meta
    charSet="utf-8"
    data-rh=""
  />,
]
`;

exports[`renders nothing and adds tags to headTags context array setting \`data-rh\` to \`null\` on title 1`] = `"<div>Yes render</div>"`;

exports[`renders nothing and adds tags to headTags context array setting \`data-rh\` to \`null\` on title 2`] = `
Array [
  <title
    data-rh={null}
  >
    Title
  </title>,
  <style
    data-rh=""
  >
    body {}
  </style>,
  <link
    data-rh=""
    href="index.css"
  />,
  <meta
    charSet="utf-8"
    data-rh=""
  />,
]
`;

exports[`renders only last meta with the same property 1`] = `
Array [
  <meta
    data-rh=""
    property="name1"
  >
    Meta 2
  </meta>,
  <meta
    data-rh=""
    name="name2"
  >
    Meta 4
  </meta>,
  <meta
    data-rh=""
    property="name3"
  >
    Meta 6
  </meta>,
]
`;

exports[`renders only the last meta with the same name 1`] = `
Array [
  <meta
    data-rh=""
  >
    Meta 2
  </meta>,
  <meta
    data-rh=""
    name="1"
  >
    Meta 3
  </meta>,
  <meta
    data-rh=""
    name="3"
  >
    Meta 5
  </meta>,
  <meta
    data-rh=""
    name="2"
  >
    Meta 6
  </meta>,
]
`;

exports[`renders only the last title 1`] = `
Array [
  <title
    data-rh=""
  >
    Title 3
  </title>,
]
`;
